ui/framework/src/ui-framework/imodel-components/models-tree/ModelsVisibilityHandler.ts

```diff
@@ -51,23 +53,25 @@ export class ModelsVisibilityHandler implements IVisibilityHandler {
   private _pendingVisibilityChange: any | undefined;
   private _subjectModelIdsCache: SubjectModelIdsCache;
   private _filteredDataProvider?: IFilteredPresentationTreeDataProvider;
+  private _elementIdsCache: ElementIdsCache;
+  private _listeners = new Array<() => void>();
 
   constructor(props: ModelsVisibilityHandlerProps) {
     this._props = props;
     this._subjectModelIdsCache = new SubjectModelIdsCache(this._props.viewport.iModel);
-    this._props.viewport.onViewedCategoriesPerModelChanged.addListener(this.onViewChanged);
-    this._props.viewport.onViewedCategoriesChanged.addListener(this.onViewChanged);
-    this._props.viewport.onViewedModelsChanged.addListener(this.onViewChanged);
-    this._props.viewport.onAlwaysDrawnChanged.addListener(this.onElementAlwaysDrawnChanged);
-    this._props.viewport.onNeverDrawnChanged.addListener(this.onElementNeverDrawnChanged);
+    this._elementIdsCache = new ElementIdsCache(this._props.viewport.iModel, this._props.rulesetId);
+    this._listeners.push(this._props.viewport.onViewedCategoriesPerModelChanged.addListener(this.onViewChanged));
+    this._listeners.push(this._props.viewport.onViewedCategoriesChanged.addListener(this.onViewChanged));
+    this._listeners.push(this._props.viewport.onViewedModelsChanged.addListener(this.onViewChanged));
+    this._listeners.push(this._props.viewport.onAlwaysDrawnChanged.addListener(this.onElementAlwaysDrawnChanged));
+    this._listeners.push(this._props.viewport.onNeverDrawnChanged.addListener(this.onElementNeverDrawnChanged));
+    if (this._props.hierarchyAutoUpdateEnabled) {
+      this._listeners.push(Presentation.presentation.onIModelHierarchyChanged.addListener(/* istanbul ignore next */() => this._elementIdsCache.clear()));

```

Comment:
I think adding a test to ensure the handler unsubscribes from these events would be useful

Reply:
Added test for unsubscribing